streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreTest.java

```diff
@@ -471,15 +475,195 @@ public void shouldDistinguishEmptyAndNull() {
+                                            Arrays.asList(SEGMENT_INTERVAL - 5, SEGMENT_INTERVAL - 10),
+                                            Arrays.asList(PUT_RETURN_CODE_VALID_TO_UNDEFINED, SEGMENT_INTERVAL - 5));
+
+        // return one values for the query with time fromTimeStamp = toTimestamp
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 4, SEGMENT_INTERVAL - 4, ResultOrder.ANY,
+                                            Collections.singletonList("v4"),
+                                            Collections.singletonList(SEGMENT_INTERVAL - 5),
+                                            Collections.singletonList(PUT_RETURN_CODE_VALID_TO_UNDEFINED));
+
+        // return values before insertion of any tombstone
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 31, SEGMENT_INTERVAL - 21, ResultOrder.ANY,
+                                            Arrays.asList("v2", "v1"),
+                                            Arrays.asList(SEGMENT_INTERVAL - 25, SEGMENT_INTERVAL - 30),
+                                            Arrays.asList(SEGMENT_INTERVAL - 20, SEGMENT_INTERVAL - 25));
+
+        // return values for the query with time range that covers both tombstones
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 24, SEGMENT_INTERVAL - 11, ResultOrder.ANY,
+                                           Collections.singletonList("v2"),
+                                           Collections.singletonList(SEGMENT_INTERVAL - 25),
+                                           Collections.singletonList(SEGMENT_INTERVAL - 20));
+
+        // return values for the query with time range that after insertion of tombstones
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 11, SEGMENT_INTERVAL - 4, ResultOrder.ANY,
+                                           Arrays.asList("v4", "v3"),
+                                           Arrays.asList(SEGMENT_INTERVAL - 5, SEGMENT_INTERVAL - 10),
+                                           Arrays.asList(PUT_RETURN_CODE_VALID_TO_UNDEFINED, SEGMENT_INTERVAL - 5));
+
+        // return all the records that are valid during the query time range but inserted beforehand
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 26, SEGMENT_INTERVAL - 5, ResultOrder.ANY,
+                                            Arrays.asList("v4", "v3", "v2", "v1"),
+                                            Arrays.asList(SEGMENT_INTERVAL - 5, SEGMENT_INTERVAL - 10, SEGMENT_INTERVAL - 25, SEGMENT_INTERVAL - 30),
+                                            Arrays.asList(PUT_RETURN_CODE_VALID_TO_UNDEFINED, SEGMENT_INTERVAL - 5, SEGMENT_INTERVAL - 20, SEGMENT_INTERVAL - 25));
+
+        // return the valid record that is still valid till the beginning of query time range (validTo = query lower time bound)
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 20, SEGMENT_INTERVAL - 15, ResultOrder.ANY,
```

Comment:
This should be `-19` to `-14`

I stop looking int the rest of the cases -- seems there is some "off by one" question?